lib/linux_admin/ssh_agent.rb

```diff
@@ -0,0 +1,85 @@
+require 'net/ssh'
+require 'fileutils'
+module LinuxAdmin
+  class SshAgent
+    attr_reader :pid
+    attr_reader :sock
+
+    def initialize(agent_socket, ssh_private_key)
+      FileUtils.mkdir_p(File.dirname(agent_socket))
+      agent_details = `ssh-agent -a #{agent_socket}`
+      @sock         = agent_details.split("=")[1].split(" ")[0].chop
+      @pid          = agent_details.split("=")[2].split(" ")[0].chop
+      IO.popen({"SSH_AUTH_SOCK" => @sock, "SSH_AGENT_PID" => @pid}, ["ssh-add", "-"], :mode => 'w') do |f|
+        f.puts(ssh_private_key)
+        if $?.to_i != 0
+          raise StandardError, "Couldn't add key to agent"
+        end
+      end
+    end
+
+    def perform_commands(ip, username, commands, _additional_flags = "")
+      result = nil
+      Net::SSH.start(ip, username, :paranoid => false, :forward_agent => true, :agent_socket_factory => -> { UNIXSocket.open(@sock) }) do |ssh|
+        commands.each do |cmd|
+          result = ssh_exec!(ssh, cmd)
+          result[:last_command] = cmd
+          break if result[:exit_status] != 0
+        end
+      end
+      result
+    end
+
+    def check_connection(ip, username, sub_ips, additional_flags)
+      connection_success = true
+      unreachable_hosts  = []
+      Net::SSH.start(ip, username, :paranoid => false, :forward_agent => true, :number_of_password_prompts => 0, :agent_socket_factory => -> { UNIXSocket.open(@sock) }) do |ssh|
+        sub_ips.each do |sub_ip|
+          result = ssh.exec!("ssh #{additional_flags} #{username}@#{sub_ip} echo $?")
+          unless result.include? "0\n"
+            connection_success = false
+            unreachable_hosts << host
+          end
+        end
+      end
+      [connection_success, unreachable_hosts]
+    end
+
+    private
+
+    def ssh_exec!(ssh, command)
```

Comment:
I think i rather keep the current version, I find it more descriptive, plus i need the request_pty, which I'm not sure we can add that way.